Strange datatype issues after PR #358

<!--- Provide a general summary of the issue in the Title above -->
Weird datatype issues associated with smol.util.container classes appear after PR #358 .

<!--- Thanks for tracking down some smelly bug!
Please fill in the following as best you can, if you have no info
for a particular section then just remove the title--->

## Current Behavior
<!--- Tell us what happens instead of the expected behavior -->
(1) When trying to test test_cofe/test_cluster_utils.py, the following issues pops up on my local computer (but not on CI):

```test_utils\test_cluster_utils.py:11 (test_int_container[IntArray1DContainer-1])
IntArrayContainer = <class 'smol.utils.cluster.container.IntArray1DContainer'>
dim = 1, rng = Generator(PCG64) at 0x25C6BA67660
​
    @pytest.mark.parametrize(
        "IntArrayContainer, dim", [(IntArray1DContainer, 1), (IntArray2DContainer, 2)]
    )
    def test_int_container(IntArrayContainer, dim, rng):
        arrays = tuple(
            rng.integers(1, 5, size=dim * (rng.integers(1, 5),)) for _ in range(10)
        )
>       container = IntArrayContainer(arrays)
​
test_cluster_utils.py:19: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
smol\utils\cluster\container.pyx:230: in smol.utils.cluster.container.IntArray1DContainer.__cinit__
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
​
>   ???
E   ValueError: Buffer dtype mismatch, expected 'long' but got 'long long'
​
smol\utils\cluster\container.pyx:251: ValueError
```


The tests were performed on a 64bit Windows 11 platform, with numpy=1.24.4.

The datatype generated by my random number generator was "long long" (int64, maybe?) rather than "long"(probably int32?). This could require an explicit specification in either the container class or the test to make the containers always store one of the types.

(2) When trying to initialize ClusterExpansion with a coef array that contains int, the following error pops up:

```self = <[UFuncTypeError(<ufunc 'divide'>, 'same_kind', dtype('float64'), dtype('int32'), 2) raised in repr()] ClusterExpansion object at 0x1856f0558b0>
    @cached_property
    def eci(self):
        """Get the ECI for the cluster expansion.
    
        This just divides coefficients by the corresponding multiplicities.
        External terms are dropped since their fitted coefficients do not
        represent ECI.
        """
        num_ext_terms = len(self._subspace.external_terms)  # check for extra terms
        coefs = self.coefs[:-num_ext_terms] if num_ext_terms else self.coefs[:]
        eci = coefs.copy()
>       eci /= self._subspace.function_total_multiplicities
E       numpy.core._exceptions._UFuncOutputCastingError: Cannot cast ufunc 'divide' output from dtype('float64') to dtype('int32') with casting rule 'same_kind'
..\venv\lib\site-packages\smol\cofe\expansion.py:182: UFuncTypeError
```

This can be temporarily solved by force transforming coefs array to float type, or do as suggested here:
[Discussion on stackoverflow](https://stackoverflow.com/questions/38673531/numpy-cannot-cast-ufunc-multiply-output-from-dtype
)

--- a/smol/cofe/expansion.py
+++ b/smol/cofe/expansion.py
@@ -179,7 +179,7 @@ class ClusterExpansion(MSONable):
         num_ext_terms = len(self._subspace.external_terms)  # check for extra terms
         coefs = self.coefs[:-num_ext_terms] if num_ext_terms else self.coefs[:]
         eci = coefs.copy()
-        eci /= self._subspace.function_total_multiplicities
+        eci = eci / self._subspace.function_total_multiplicities
         return eci
 
     @cached_property

--- a/smol/cofe/wrangling/wrangler.py
+++ b/smol/cofe/wrangling/wrangler.py
@@ -426,7 +426,7 @@ class StructureWrangler(MSONable):
                 ) from error
 
         if normalize:
-            properties /= self.sizes
+            properties = properties / self.sizes
 
         return properties
 

--- a/tests/test_utils/test_cluster_utils.py
+++ b/tests/test_utils/test_cluster_utils.py
@@ -1,3 +1,4 @@
+import numpy as np
 import pytest
 
 from smol.utils.cluster.container import (
@@ -13,15 +14,18 @@ from smol.utils.cluster.container import (
     "IntArrayContainer, dim", [(IntArray1DContainer, 1), (IntArray2DContainer, 2)]
 )
 def test_int_container(IntArrayContainer, dim, rng):
+    # Use np.int_ to ensure cross-platform compatibility.
     arrays = tuple(
-        rng.integers(1, 5, size=dim * (rng.integers(1, 5),)) for _ in range(10)
+        rng.integers(1, 5, size=dim * (rng.integers(1, 5),)).astype(np.int_)
+        for _ in range(10)
     )
     container = IntArrayContainer(arrays)
     assert len(container) == 10
 
     # test setting the same sized list (no memory reallocation)
     new_arrays = tuple(
-        rng.integers(1, 5, size=dim * (rng.integers(1, 5),)) for _ in range(10)
+        rng.integers(1, 5, size=dim * (rng.integers(1, 5),)).astype(np.int_)
+        for _ in range(10)
     )
     container.set_arrays(new_arrays)
 
@@ -29,7 +33,8 @@ def test_int_container(IntArrayContainer, dim, rng):
 
     # test setting a different sized list (needs memory reallocation)
     new_arrays = tuple(
-        rng.integers(1, 5, size=dim * (rng.integers(1, 5),)) for _ in range(12)
+        rng.integers(1, 5, size=dim * (rng.integers(1, 5),)).astype(np.int_)
+        for _ in range(12)
     )
     container.set_arrays(new_arrays)
 
@@ -37,7 +42,7 @@ def test_int_container(IntArrayContainer, dim, rng):
 
     with pytest.raises(ValueError):
         new_arrays = tuple(
-            rng.integers(1, 5, size=(dim + 1) * (rng.integers(1, 5),))
+            rng.integers(1, 5, size=(dim + 1) * (rng.integers(1, 5),)).astype(np.int_)
             for _ in range(12)
         )
         container.set_arrays(new_arrays)
@@ -79,7 +84,9 @@ def test_orbit_container(rng):
         orbit_id = i
         bit_id = rng.integers(1, 5, dtype=int)
         correlation_tensors = rng.random(tuple(rng.integers(1, 5, size=2)))
-        tensor_indices = rng.integers(1, 5, size=correlation_tensors.shape[0])
+        tensor_indices = rng.integers(1, 5, size=correlation_tensors.shape[0]).astype(
+            np.int_
+        )
         orbit_data.append((orbit_id, bit_id, correlation_tensors, tensor_indices))
 
     container = OrbitContainer(tuple(orbit_data))
@@ -91,7 +98,9 @@ def test_orbit_container(rng):
         orbit_id = i
         bit_id = rng.integers(1, 5, dtype=int)
         correlation_tensors = rng.random(tuple(rng.integers(1, 5, size=2)))
-        tensor_indices = rng.integers(1, 5, size=correlation_tensors.shape[0])
+        tensor_indices = rng.integers(1, 5, size=correlation_tensors.shape[0]).astype(
+            np.int_
+        )
         new_orbit_data.append((orbit_id, bit_id, correlation_tensors, tensor_indices))
 
     container.set_orbits(tuple(new_orbit_data))
@@ -103,7 +112,9 @@ def test_orbit_container(rng):
         orbit_id = i
         bit_id = rng.integers(1, 5, dtype=int)
         correlation_tensors = rng.random(tuple(rng.integers(1, 5, size=2)))
-        tensor_indices = rng.integers(1, 5, size=correlation_tensors.shape[0])
+        tensor_indices = rng.integers(1, 5, size=correlation_tensors.shape[0]).astype(
+            np.int_
+        )
         new_orbit_data.append((orbit_id, bit_id, correlation_tensors, tensor_indices))
 
     container.set_orbits(tuple(new_orbit_data))

